### DISCUSSION TITLE
Why must the glow painter be destroyed manually

### DISCUSSION THREAD

**Discussion by DusterTheFirst** [emilk/egui#1386] [Q&A]
The following message is printed by tracing after the egui application panics:

```
WARN egui_glow::painter: You forgot to call destroy() on the egui glow painter. Resources will leak!
```

Tracking down this message it leads me to the `egui_glow` painter. In the documentation comment, it notes:

https://github.com/emilk/egui/blob/465c96122cf6aea72e7fbd8d6e7c2c048e88424f/egui_glow/src/painter.rs#L27-L29

I was wondering what the rationale behind not just using a `Drop` implementation to clean up the painter resources. There even is a `Drop` handler that ensures the programmer called this destroy function.

https://github.com/emilk/egui/blob/465c96122cf6aea72e7fbd8d6e7c2c048e88424f/egui_glow/src/painter.rs#L630-L638

To me, it would make more sense to call `Painter::destroy` in the `Drop` implementation. So that the painter is always cleaned up without the programmer needing to manually call `Painter::destroy`, and it would `destroy` the painter in the case of a panic

**Comment by emilk** [comment]
The problem is that `destroy` uses the `glow::Context`, and must be called when the context is still valid, and I think it will be invalidated when the window it is attached to is destroyed. If we called `destroy` from `Drop` this could easily happen, and could lead to UB (since all `glow` functions are `unsafe`).

**Reply by DusterTheFirst** [reply]
Would it be possible to check if the context is valid before calling destroy on it?

**Comment by CompeyDev** [comment]
Right, I'm using a `RetainedImage`, and I seem to be getting the same warning regarding the glow painter not being destroyed, even though I am not directly using it.

How shall I destroy a glow painter in this case?

**Comment by xairaven** [comment]
I'm also interested in the solution. @emilk, could you shed some light on this issue?
